File: client/ios/Data/Support/SupportHttpGateway.swift
import Foundation

public protocol SupportGateway: Sendable {
    func load(path: String) async throws -> SupportPagePayload
    func submit(path: String, fields: [String: String]) async throws -> SupportPagePayload
}

public struct SupportHttpGateway: SupportGateway {
    private let baseUrl: String
    private let session: URLSession

    public init(baseUrl: String, session: URLSession = .shared) {
        self.baseUrl = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.session = session
    }

    public func load(path: String) async throws -> SupportPagePayload {
        try await request(path: path, method: "GET", fields: nil)
    }

    public func submit(path: String, fields: [String: String]) async throws -> SupportPagePayload {
        try await request(path: path, method: "POST", fields: fields)
    }

    private func request(path: String, method: String, fields: [String: String]?) async throws -> SupportPagePayload {
        guard path.hasPrefix("/support") else { throw SupportGatewayError.invalidPath }
        guard let url = URL(string: baseUrl + path) else { throw SupportGatewayError.invalidUrl }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let fields {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: fields)
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw SupportGatewayError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
            throw SupportGatewayError.requestFailed(object?["message"] as? String ?? "Support request failed with HTTP \(http.statusCode).")
        }
        let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        return parse(root)
    }

    private func parse(_ root: [String: Any]) -> SupportPagePayload {
        let data = root["data"] as? [String: Any] ?? [:]
        let meta = root["meta"] as? [String: Any] ?? [:]
        let locations = (root["interface"] as? [String: Any])?["locations"] as? [String: Any]
        let content = (locations?["shell.main.content"] as? [[String: Any]])?.first
        let rowDictionaries = data["rows"] as? [[String: Any]] ?? []

        let rows = rowDictionaries.compactMap { row -> SupportRowPayload? in
            guard row["availableItems"] != nil || row["request"] != nil else { return nil }
            return SupportRowPayload(
                id: string(row["id"]),
                context: string(row["context"]),
                request: string(row["request"]),
                description: string(row["description"]),
                href: string(row["href"]),
                availableItems: integer(row["availableItems"])
            )
        }
        let cases = rowDictionaries.compactMap { row -> CaseRowPayload? in
            guard row["reference"] != nil, row["availableItems"] == nil else { return nil }
            return CaseRowPayload(
                reference: string(row["reference"]),
                context: string(row["context"]),
                category: string(row["category"]),
                status: string(row["status"]),
                href: string(row["href"])
            )
        }

        let informationForm = data["informationForm"] as? [String: Any]
        let fieldDictionaries = data["formFields"] as? [[String: Any]]
            ?? informationForm?["fields"] as? [[String: Any]]
            ?? []
        let fields = fieldDictionaries.map { field in
            SupportFieldPayload(
                name: string(field["nameEntity"]),
                label: string(field["label"]),
                type: string(field["type"], fallback: "text"),
                value: optionalString(field["value"]),
                required: boolean(field["required"]),
                options: (field["options"] as? [[String: Any]] ?? []).map {
                    SupportOptionPayload(label: string($0["label"]), value: string($0["value"]))
                }
            )
        }
        let actions = (data["headerActions"] as? [[String: Any]] ?? []).map {
            SupportActionPayload(
                label: string($0["label"]),
                href: string($0["href"]),
                method: string($0["method"], fallback: "GET"),
                enabled: ($0["enabled"] as? Bool) ?? true
            )
        }

        return SupportPagePayload(
            title: string(meta["title"], fallback: string(content?["label"], fallback: "Support")),
            description: string(content?["description"]),
            rows: rows,
            cases: cases,
            fields: fields,
            actions: actions,
            action: optionalString(data["action"]) ?? optionalString(informationForm?["action"]),
            method: string(data["method"], fallback: string(informationForm?["method"], fallback: "GET")),
            reference: optionalString(data["reference"]) ?? optionalString(data["caseReference"]),
            status: optionalString(data["status"]),
            businessContext: optionalString(data["businessContext"]),
            category: optionalString(data["category"]),
            descriptionText: optionalString(data["description"])
        )
    }

    private func string(_ value: Any?, fallback: String = "") -> String {
        if let value = value as? String { return value }
        if let value { return String(describing: value) }
        return fallback
    }

    private func optionalString(_ value: Any?) -> String? {
        let value = string(value)
        return value.isEmpty ? nil : value
    }

    private func integer(_ value: Any?) -> Int {
        if let value = value as? Int { return value }
        if let value = value as? NSNumber { return value.intValue }
        return Int(string(value)) ?? 0
    }

    private func boolean(_ value: Any?) -> Bool {
        if let value = value as? Bool { return value }
        if let value = value as? NSNumber { return value.boolValue }
        return ["1", "true", "yes"].contains(string(value).lowercased())
    }
}

private enum SupportGatewayError: LocalizedError {
    case invalidPath
    case invalidUrl
    case invalidResponse
    case requestFailed(String)

    var errorDescription: String? {
        switch self {
        case .invalidPath: return "Support path is invalid."
        case .invalidUrl: return "Support gateway URL is invalid."
        case .invalidResponse: return "Support gateway returned an invalid response."
        case .requestFailed(let message): return message
        }
    }
}

File: client/android/Data/Support/SupportHttpGateway.kt
package app.mobiling.client.data.support

import app.mobiling.client.contract.support.CaseRowPayload
import app.mobiling.client.contract.support.SupportActionPayload
import app.mobiling.client.contract.support.SupportFieldPayload
import app.mobiling.client.contract.support.SupportOptionPayload
import app.mobiling.client.contract.support.SupportPagePayload
import app.mobiling.client.contract.support.SupportRowPayload
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

interface SupportGateway {
    suspend fun load(path: String): SupportPagePayload
    suspend fun submit(path: String, fields: Map<String, String>): SupportPagePayload
}

class SupportHttpGateway(
    private val baseUrl: String,
    private val client: OkHttpClient = OkHttpClient(),
) : SupportGateway {
    private val jsonMediaType = "application/json".toMediaType()

    override suspend fun load(path: String): SupportPagePayload = request("GET", path, null)

    override suspend fun submit(path: String, fields: Map<String, String>): SupportPagePayload =
        request("POST", path, JSONObject(fields))

    private suspend fun request(method: String, path: String, body: JSONObject?): SupportPagePayload = withContext(Dispatchers.IO) {
        require(path.startsWith("/support")) { "Casing support path must stay under /support." }
        val builder = Request.Builder()
            .url(baseUrl.trimEnd('/') + path)
            .header("Accept", "application/json")
        if (method == "POST") {
            builder.post((body ?: JSONObject()).toString().toRequestBody(jsonMediaType))
        } else {
            builder.get()
        }
        client.newCall(builder.build()).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                val message = runCatching { JSONObject(responseBody).optString("message") }.getOrNull().orEmpty()
                error(message.ifBlank { "Support request failed with HTTP ${response.code}." })
            }
            parse(JSONObject(responseBody.ifBlank { "{}" }))
        }
    }

    private fun parse(root: JSONObject): SupportPagePayload {
        val data = root.optJSONObject("data") ?: JSONObject()
        val meta = root.optJSONObject("meta") ?: JSONObject()
        val content = root.optJSONObject("interface")
            ?.optJSONObject("locations")
            ?.optJSONArray("shell.main.content")
            ?.optJSONObject(0)
        val rows = data.optJSONArray("rows") ?: JSONArray()
        val supportRows = mutableListOf<SupportRowPayload>()
        val caseRows = mutableListOf<CaseRowPayload>()
        for (index in 0 until rows.length()) {
            val row = rows.optJSONObject(index) ?: continue
            if (row.has("availableItems") || row.has("request")) {
                supportRows += SupportRowPayload(
                    id = row.optString("id"),
                    context = row.optString("context"),
                    request = row.optString("request"),
                    description = row.optString("description"),
                    href = row.optString("href"),
                    availableItems = row.optInt("availableItems", 0),
                )
            } else if (row.has("reference")) {
                caseRows += CaseRowPayload(
                    reference = row.optString("reference"),
                    context = row.optString("context"),
                    category = row.optString("category"),
                    status = row.optString("status"),
                    href = row.optString("href"),
                )
            }
        }

        return SupportPagePayload(
            title = meta.optString("title", content?.optString("label", "Support") ?: "Support"),
            description = content?.optString("description").orEmpty(),
            rows = supportRows,
            cases = caseRows,
            fields = parseFields(data.optJSONArray("formFields") ?: data.optJSONObject("informationForm")?.optJSONArray("fields")),
            actions = parseActions(data.optJSONArray("headerActions") ?: data.optJSONArray("actions")),
            action = data.optString("action").takeIf(String::isNotBlank)
                ?: data.optJSONObject("informationForm")?.optString("action")?.takeIf(String::isNotBlank),
            method = data.optString("method", data.optJSONObject("informationForm")?.optString("method", "GET") ?: "GET"),
            reference = data.optString("reference").takeIf(String::isNotBlank)
                ?: data.optString("caseReference").takeIf(String::isNotBlank),
            status = data.optString("status").takeIf(String::isNotBlank),
            businessContext = data.optString("businessContext").takeIf(String::isNotBlank),
            category = data.optString("category").takeIf(String::isNotBlank),
            descriptionText = data.optString("description").takeIf(String::isNotBlank),
            informationQuestion = data.optJSONObject("informationRequest")?.optString("question")?.takeIf(String::isNotBlank),
        )
    }

    private fun parseFields(array: JSONArray?): List<SupportFieldPayload> = buildList {
        if (array == null) return@buildList
        for (index in 0 until array.length()) {
            val field = array.optJSONObject(index) ?: continue
            val options = field.optJSONArray("options") ?: JSONArray()
            add(SupportFieldPayload(
                name = field.optString("nameEntity"),
                label = field.optString("label"),
                type = field.optString("type", "text"),
                value = field.opt("value")?.toString()?.takeIf { it != "null" },
                required = field.optBoolean("required", false),
                options = buildList {
                    for (optionIndex in 0 until options.length()) {
                        val option = options.optJSONObject(optionIndex) ?: continue
                        add(SupportOptionPayload(option.optString("label"), option.optString("value")))
                    }
                },
            ))
        }
    }

    private fun parseActions(array: JSONArray?): List<SupportActionPayload> = buildList {
        if (array == null) return@buildList
        for (index in 0 until array.length()) {
            val action = array.optJSONObject(index) ?: continue
            add(SupportActionPayload(
                label = action.optString("label"),
                href = action.optString("href"),
                method = action.optString("method", "GET"),
                enabled = action.optBoolean("enabled", true),
            ))
        }
    }
}
